cpp/src/parquet/encoding.cc

```diff
@@ -2236,6 +2326,78 @@ class DeltaByteArrayDecoder : public DecoderImpl,
   ByteArray last_value_;
 };
 
+// ----------------------------------------------------------------------
+// BYTE_STREAM_SPLIT
+
+template <typename DType>
+class ByteStreamSplitDecoder : public DecoderImpl, virtual public TypedDecoder<DType> {
+ public:
+  using T = typename DType::c_type;
+  explicit ByteStreamSplitDecoder(const ColumnDescriptor* descr);
+
+  int Decode(T* buffer, int max_values) override;
+
+  int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                  int64_t valid_bits_offset,
+                  typename EncodingTraits<DType>::Accumulator* builder) override;
+
+  int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                  int64_t valid_bits_offset,
+                  typename EncodingTraits<DType>::DictAccumulator* builder) override;
+
+  void SetData(int num_values, const uint8_t* data, int len) override;
+
+ private:
+  int num_values_{0U};
+  int num_values_decoded_{0U};
+};
+
+template <typename DType>
+ByteStreamSplitDecoder<DType>::ByteStreamSplitDecoder(const ColumnDescriptor* descr)
+    : DecoderImpl(descr, Encoding::BYTE_STREAM_SPLIT) {}
+
+template <typename DType>
+void ByteStreamSplitDecoder<DType>::SetData(int num_values, const uint8_t* data,
+                                            int len) {
+  DecoderImpl::SetData(num_values, data, len);
+  num_values_ = num_values;
+  num_values_decoded_ = 0;
+}
+
+template <typename DType>
+int ByteStreamSplitDecoder<DType>::Decode(T* buffer, int max_values) {
+  constexpr size_t num_streams = sizeof(T);
+  using UnsignedType = typename UnsignedTypeWithWidth<sizeof(T)>::type;
+  const int values_to_decode = std::min(num_values_ - num_values_decoded_, max_values);
+  for (int i = 0; i < values_to_decode; ++i) {
+    UnsignedType reconstructed_value_as_uint{0U};
+    for (size_t b = 0; b < num_streams; ++b) {
+      const size_t byte_index = b * num_values_ + num_values_decoded_ + i;
+      const UnsignedType byte_value = static_cast<UnsignedType>(data_[byte_index]);
+      reconstructed_value_as_uint |= (byte_value << (8U * b));
+    }
+    const T reconstructed_value_as_fp =
+        *reinterpret_cast<const T*>(&reconstructed_value_as_uint);
+    buffer[i] = reconstructed_value_as_fp;
```

Comment:
Or could simply write `memcpy(buffer + i, &reconstructed_value_as_uint, sizeof(T))`

Reply:
Code was refactored so the memcpy is not present there atm.